File: vault/csharp/Naveego.Vault/Secret.cs
using System.Collections.Generic;
using Vault.Models;

namespace Metabase.Api.Vault
{
    /// <summary>
    ///     Represents the case where no data is expected back from the API.
    /// </summary>
    public class NoData
    {
    }

    public class Secret<T>
    {
        public string RequestId { get; set; }

        public string LeaseId { get; set; }

        public bool Renewable { get; set; }

        public int LeaseDuration { get; set; }

        public List<string> Warnings { get; set; }

        public T Data { get; set; }

        public SecretAuth Auth { get; set; }

        internal static Secret<T> FromResponse(VaultResponse<T> response)
        {
            if (response == null)
            {
                return null;
            }

            return new Secret<T>
            {
                Data = response.Data,
                Auth = response.Auth == null
                    ? null
                    : new SecretAuth
                    {
                        Accessor = response.Auth.Accessor,
                        Metadata = response.Auth.Metadata,
                        Policies = response.Auth.Policies,
                        Renewable = response.Auth.Renewable,
                        ClientToken = response.Auth.ClientToken,
                        LeaseDuration = response.Auth.LeaseDuration
                    },
                Renewable = response.Renewable,
                LeaseId = response.LeaseId,
                LeaseDuration = response.LeaseDuration,
                Warnings = response.Warnings,
                RequestId = response.RequestId
            };
        }

        /// <summary>
        ///     Returns a new secret with the new <paramref name="data" />
        ///     and a shallow copy of all other properties.
        /// </summary>
        /// <param name="data"></param>
        /// <typeparam name="TOther"></typeparam>
        /// <returns></returns>
        public Secret<TOther> ReWrap<TOther>(TOther data)
        {
            return new Secret<TOther>
            {
                Data = data,
                Renewable = Renewable,
                LeaseDuration = LeaseDuration,
                Auth = Auth,
                Warnings = Warnings,
                LeaseId = LeaseId,
                RequestId = RequestId
            };
        }
    }


    public class SecretAuth
    {
        public string ClientToken { get; set; }
        public string Accessor { get; set; }
        public List<string> Policies { get; set; }
        public Dictionary<string, string> Metadata { get; set; }
        public int LeaseDuration { get; set; }
        public bool Renewable { get; set; }
    }
}

File: vault/java/src/main/scala/com/naveego/vault/Secret.scala
package com.naveego.vault

import spray.json._


case class Secret[T](
                   data: Option[T],
                   auth: SecretAuth = null,
                   renewable: Boolean = false,
                   leaseDuration: Long = 0,
                   leaseId: String = null,
                   warnings: List[String] = null,
                   requestId: String = ""
                 ) {
  /**
    * Returns a new Secret with `newData` as the data.
    * @param newData
    * @tparam TOther
    * @return
    */
  def rewrap[TOther:JsonFormat](newData: TOther): Secret[TOther] = {
    this.copy[TOther](data = Some(newData))
  }


  implicit class JsonSecret(self: Secret[JsObject]) {
    /**
      * Returns a secret with the same field values but the data parsed
      * into TFormat.
      * @tparam TFormat
      * @return
      */
    def withParsedValue[TFormat: JsonFormat]() : Secret[TFormat] = {
      self.data match {
        case Some(j) => self.rewrap[TFormat](j.convertTo[TFormat])
        case None => self.copy[TFormat](data = None)
      }
    }
  }
}


case class SecretAuth(
                       clientToken: String,
                       renewable: Boolean = false,
                       leaseDuration: Long = 0,
                       policies: List[String] = List[String](),
                       metadata: Map[String, String] = null,
                       accessor: String = null
                     )

object NaveegoJsonProtocol extends DefaultJsonProtocol {


  implicit val secretAuth: RootJsonFormat[SecretAuth] = new RootJsonFormat[SecretAuth] {
    override def write(obj: SecretAuth): JsValue = {
      JsObject(
        "client_token" -> obj.clientToken.toJson,
        "accessor" -> obj.accessor.toJson,
        "policies" -> obj.policies.toJson,
        "metadata" -> obj.metadata.toJson,
        "lease_duration" -> obj.leaseDuration.toJson,
        "renewable" -> obj.renewable.toJson
      )
    }

    override def read(json: JsValue): SecretAuth = {
      if (json == null || json == JsNull) {
        return null
      }

      val obj = json.asJsObject

      new SecretAuth(
        obj.fields("client_token").convertTo[String],
        obj.fields("renewable").convertTo[Boolean],
        obj.fields("lease_duration").convertTo[Int],
        obj.fields("policies").convertTo[List[String]],
        obj.fields("metadata") match {
          case o: JsObject => o.convertTo[Map[String, String]]
          case _ => null
        },
        obj.fields("accessor").convertTo[String]
      )
    }
  }

  implicit def secret[T: JsonFormat] = new RootJsonFormat[Secret[T]] {
    override def write(obj: Secret[T]): JsValue = {
      JsObject(
        "data" -> obj.data.toJson,
        "auth" -> obj.auth.toJson,
        "request_id" -> obj.requestId.toJson,
        "lease_id" -> obj.leaseId.toJson,
        "lease_duration" -> obj.leaseDuration.toJson,
        "renewable" -> obj.renewable.toJson,
        "warnings" -> obj.warnings.toJson
      )
    }

    override def read(json: JsValue): Secret[T] = {
      val obj = json.asJsObject

      new Secret(
        obj.fields("data") match {
          case o: JsObject => Some(o.convertTo[T])
          case _ => None
        },
        obj.fields("auth") match {
          case o: JsObject => o.convertTo[SecretAuth]
          case _ => null
        },
        obj.fields("renewable").convertTo[Boolean],
        obj.fields("lease_duration").convertTo[Int],
        obj.fields("lease_id").convertTo[String],
        obj.fields("warnings") match {
          case o: JsArray => o.convertTo[List[String]]
          case _ => null
        },
        obj.fields("request_id").convertTo[String]
      )
    }
  }


}
